src/fsharp/block.fs

```diff
@@ -192,25 +160,22 @@ module Block =
         for i = 0 to arr.Length - 1 do
             if predicate arr.[i] then
                 builder.Add(arr.[i])
-        builder.ToImmutable()
+        builder.Capacity <- builder.Count
+        builder.MoveToImmutable()
 
     let exists predicate (arr: block<'T>) =
-        let n = arr.Length
-        let mutable result = false
-        let mutable i = 0
-        while not result && i < n do
-            if predicate arr.[i] then
-                result <- true
-            i <- i + 1
-        result
+        let len = arr.Length
+        let rec loop i = i < len && (predicate arr.[i] || loop (i+1))
+        len > 0 && loop 0
 
     let choose (chooser: 'T -> 'U option) (arr: block<'T>) : block<'U> =
         let builder = ImmutableArray.CreateBuilder(arr.Length)
         for i = 0 to arr.Length - 1 do
             let result = chooser arr.[i]
             if result.IsSome then
                 builder.Add(result.Value)
-        builder.ToImmutable()
+        builder.Capacity <- builder.Count
```

Comment:
If you don't know the length it doesn't matter too much if you do ToImmutable. You'll do a copy by changing capacity anyway.

Reply:
Changing the capacity will only copy if the count is different.

Reply:
Fair enough :)